pymtl3/stdlib/ifcs/CalleeIfc.py

```diff
@@ -0,0 +1,167 @@
+# CalleeIfc.py
+#=========================================================================
+# RTL method interface. RTL equivalence of NonBlockingCalleeIfc in CL
+#
+# Author : Yixiao Zhang
+#   Date : June 10, 2019
+
+from __future__ import absolute_import, division, print_function
+
+from copy import deepcopy
+
+from pymtl3 import *
+from pymtl3.dsl.errors import InvalidConnectionError
+
+
+#-------------------------------------------------------------------------
+# CalleeRTL2CL
+#-------------------------------------------------------------------------
+class CalleeRTL2CL( Component ):
+
+  def construct( s, ifc_rtl_callee ):
+
+    ifc_rtl_caller = deepcopy( ifc_rtl_callee )
+    ifc_rtl_caller._dsl.constructed = False
+    s.ifc_rtl_caller = ifc_rtl_caller.inverse()
+
+    s.called = False
+    s.rdy = False
+
+    # generate upblk depending on args
+    if s.ifc_rtl_caller.args:
+      # has args, create tmp var
+      s.args = s.ifc_rtl_caller.args._dsl.Type()
+
+      # update args & en
+      @s.update
+      def update_en_args():
+        s.ifc_rtl_caller.en = Bits1( 1 ) if s.called else Bits1( 0 )
+        s.ifc_rtl_caller.args = s.args
+        s.called = False
+
+      # add constraints between callee method and upblk
+      s.add_constraints( M( s.cl_callee_method ) < U( update_en_args ) )
+
+      # know which method to add constraints on later
+      cl_method = s.cl_callee_method
+
+    else:
+      # no args, update en
+      @s.update
+      def update_en():
+        s.ifc_rtl_caller.en = Bits1( 1 ) if s.called else Bits1( 0 )
+        s.called = False
+
+      # add constraints between callee method and upblk
+      s.add_constraints( M( s.cl_callee_method_no_arg ) < U( update_en ) )
+
+      # know which method to add constraints on later
+      cl_method = s.cl_callee_method_no_arg
+
+    # generate upblk depending on rets
+    if s.ifc_rtl_caller.rets:
+      # create tmp var for rets
+      s.rets = s.ifc_rtl_caller.rets._dsl.Type()
+
+      # generate upblk for rets
+      @s.update
+      def update_rets():
+        s.rets = s.ifc_rtl_caller.rets
+
+      s.add_constraints( U( update_rets ) < M( cl_method ) )
+
+    else:
+      # return None if no ret
+      s.rets = None
+
+    # Generate upblk and add constraints for rdy
+    @s.update
+    def update_rdy():
+      s.rdy = True if s.ifc_rtl_caller.rdy else False
+
+    s.add_constraints( U( update_rdy ) < M( cl_method ) )
+
+  @non_blocking( lambda s: s.rdy )
+  def cl_callee_method( s, **kwargs ):
+    s.args.__dict__.update( kwargs )
+    s.called = True
+    return s.rets
+
+  @non_blocking( lambda s: s.rdy )
+  def cl_callee_method_no_arg( s ):
+    s.called = True
+    return s.rets
+
+
+#-------------------------------------------------------------------------
+# CalleeIfcRTL
+#-------------------------------------------------------------------------
+class CalleeIfcRTL( Interface ):
+
+  def construct( s, ArgTypes=None, RetTypes=None ):
+    if ArgTypes:
```

Comment:
I think we could have a CallerIfcRTL which is basically the inverse.

Reply:
Hmm maybe we just rename the file to generic_method_ifcs.py and you implement both CalleeIfcRTL, CallerIfcRTL, and the adapter, in the same file?